### PR TITLE
Support FTPES(FTPS explicit mode)

### PR DESCRIPTION
**Pull request by sakama**
I changed code to support FTPES(FTPS explicit mode).
Current implementation works with FTPS(FTPS implicit mode), but doesn't works with FTPES(FTPS explicit mode).

I added `ssl_explicit(default: true)` option to support both of them.
I think it's better that plugin will use FTPES as a default mode. FTPES seems to be used widely than FTPS(implicit).
- FTPS(implicit) was written at [RFC4217 ver.7](https://tools.ietf.org/html/draft-murray-auth-ftp-ssl-07), but was removed at [ver.8](https://tools.ietf.org/html/draft-murray-auth-ftp-ssl-08).
- system administrator could force user to use FTPES instead of unencrypted FTP if server provides FTPES.
- vsftpd(probably a most common FTP daemon) uses FTPES as a default mode of FTPS.
## Backward compatibility

Although I changed default value of `port` option, I think this change keeps backward compatibility.
Because current version throw Exception if user executes without `port` option's setting.

### CODE REVIEW

File: src/main/java/org/embulk/input/FtpFileInputPlugin.java [hunk 1]
```diff
@@ -66,7 +66,7 @@
         public String getHost();
 
         @Config("port")
-        @ConfigDefault("null")
+        @ConfigDefault("21")
```
Comment (frsyuki) [inline src/main/java/org/embulk/input/FtpFileInputPlugin.java]: why is this change here?

Comment (sakama) [inline src/main/java/org/embulk/input/FtpFileInputPlugin.java]: I moved into [newFTPClient()](https://github.com/embulk/embulk-input-ftp/pull/5/commits/6acf091ca7d322cf8b2b5b5c13783a5ed875a018#diff-6c1d7324b8325e41503f63c5192e8ee4R182) and changed default value as follows.
- FTP 21
- FTPES 21
- FTPS 990

File: src/main/java/org/embulk/input/FtpFileInputPlugin.java [hunk 2]
```diff
@@ -68,6 +71,7 @@
         @Config("port")
         @ConfigDefault("null")
         public Optional<Integer> getPort();
+        public void setPort(Optional<Integer> port);
```
Comment (muga) [inline src/main/java/org/embulk/input/FtpFileInputPlugin.java]: This method adding is ok. But I still don't know why it is added. It seems that this method is only used for saving the value of default port within newFTPClient method. Please correct me if no.

File: src/main/java/org/embulk/input/FtpFileInputPlugin.java [hunk 3]
```diff
@@ -51,6 +51,9 @@
         implements FileInputPlugin
 {
     private final Logger log = Exec.getLogger(FtpFileInputPlugin.class);
+    private static final Integer FTP_DEFULAT_PORT = 21;
+    private static final Integer FTPS_DEFAULT_PORT = 990;
+    private static final Integer FTPES_DEFAULT_PORT = 21;
```
Comment (muga) [inline src/main/java/org/embulk/input/FtpFileInputPlugin.java]: Minor: Can you change the types with `int`?

**Review by muga — COMMENTED** [review]
I don't think that the setPort method is good in this case. Do you have special reasons?

File: src/main/java/org/embulk/input/FtpFileInputPlugin.java (hunk 3 above)
Comment (sakama) [reply src/main/java/org/embulk/input/FtpFileInputPlugin.java]: I changed to int instead of Integer. Thanks.

File: src/main/java/org/embulk/input/FtpFileInputPlugin.java (hunk 2 above)
Comment (sakama) [reply src/main/java/org/embulk/input/FtpFileInputPlugin.java]: I created 3 types default ports depends on FTP mode.
- FTP 21
- FTPS 990
- FTPES 21

I created this method to set port number from Java.
I think this method is unnecessary if there is only 1 default port type as you're saying.

``` java
int defaultPort = FTP_DEFULAT_PORT;
if (task.getSsl()) {
    ...
    if (task.getSslExplicit()) {
        ...
        defaultPort = FTPES_DEFAULT_PORT;
        ...
    }
    else {
        ...
        defaultPort = FTPS_DEFAULT_PORT;
        ...
    }
}

if (!task.getPort().isPresent()) {
    task.setPort(Optional.of(defaultPort));
}
```

Comment (muga) [reply src/main/java/org/embulk/input/FtpFileInputPlugin.java]: I mean, why we need to set (reset) the value as a config/task source? I don't think that it's not good to modify a config/task source basically except setFiles in this case. Because, port is deterministic. newFTPClient method is called anywhere, anytime. the default port is constant. (In the case of listFiles, they are not deterministic as you might know.)

Comment (sakama) [reply src/main/java/org/embulk/input/FtpFileInputPlugin.java]: Understand. I deleted setPort() method.

**Comment by muga** [comment]
Looks good to me. please merge it.
